src/clj/rems/application/process_managers.clj

```diff
@@ -0,0 +1,44 @@
+(ns rems.application.process-managers
+  "Miscellaneous process managers.
+
+  NB: An event manager should return an empty sequence (or `nil`) if it doesn't create new events itself."
+  (:require [clojure.set :refer [difference]]
+            [rems.api.services.attachment :as attachment]
+            [rems.api.services.blacklist :as blacklist]
+            [rems.common.application-util :as application-util]
+            [rems.db.applications :as applications]
+            [rems.db.attachments :as attachments]))
+
+(defn revokes-to-blacklist
+  "Revokation causes the users to be blacklisted."
+  [new-events]
+  (doseq [event new-events
+          :when (= :application.event/revoked (:event/type event))
+          :let [application (applications/get-application-internal (:application/id event))]
+          user (application-util/applicant-and-members application)
+          resource (:application/resources application)]
+    (blacklist/add-user-to-blacklist! (:event/actor event)
+                                      {:blacklist/user {:userid (:userid user)}
+                                       :blacklist/resource {:resource/ext-id (:resource/ext-id resource)}
+                                       :comment (:application/comment event)})))
+
+(defn delete-applications
+  "The deleted event causes a side-effect that completely deletes the application."
+  [new-events]
+  (doseq [event new-events]
+    (when (= :application.event/deleted (:event/type event))
+      (applications/delete-application-and-reload-cache! (:application/id event)))))
+
+(defn delete-orphan-attachments [application-id]
+  (let [application (applications/get-application-internal application-id)
+        attachments-in-use (rems.api.services.attachment/get-attachments-in-use application)
+        all-attachments (set (map :attachment/id (:application/attachments application)))]
+    (doseq [attachment-id (difference all-attachments attachments-in-use)]
+      (attachments/delete-attachment! attachment-id))))
+
+(defn delete-orphan-attachments-on-submit
+  "When an application is submitted, we delete its unused attachments, if any."
+  [new-events]
+  (doseq [event new-events]
+    (when (= :application.event/submitted (:event/type event))
```

Comment:
hypothetically remark and review can both generate attachments that can linger around, not sure if that would ever happen but if it did, then maybe `:application.event/closed` could also trigger delete orphan attachments?

Reply:
Could be most state changes actually, because a lot of them have the comment that can include an attachment. I think we have to still decide if we want to a) check this in every state change b) have a background job.

Reply:
Added this concern to https://github.com/CSCfi/rems/issues/3039

Maybe it's an easier solution than the background job.